### PR TITLE
tforkner/3556-move-budget moved budget to common

### LINKED ISSUE #3556: [Feature] Move budget calculations out of redux and refactor
**Description and related issues -**
We are currently doing all of the budget calculations on the frontend in redux. This is working while there is only one person, but it requires the budget to be calculated on each users machine every time there is a change. We should move this functionality out of redux so that it can be used on the backend and stored in the APD.

The code is also a little hard to follow. Refactor it to make it more functional and add more tests for more edge cases.

**What is the goal or purpose of this ticket?**
Prepping for multi-user experience.

### This task is done when…

- [ ] move the code to the common package, so that it can be used in the backend but also the front-end if needed
- [ ] refactor the budget so that it's as clear as possible
- [ ] write more test for the budget code

### PR DESCRIPTION
**Pull request by thetif**
Resolves #3556

### Description
Moving the budget information from the frontend to the shared package

### Significant changes or possible side effects
Possible budget issues

### Steps to manually verify this change
Make sure that nothing seems weird with the budget. I feel like there are enough budget tests that we can probably mostly rely on them to verify this change.

### This pull request is ready to review when

- [x] Automated tests are updated (and all tests are passing)
- [x] Pull request has been labeled, if applicable with feature, content, bug, tests, refactor

### This pull request can be merged when

- [x] Code has been reviewed by someone other than the original author
- [ ] QA has verified the accessibility and functionality related to the change
- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

File: common/utils/budget.js [hunk 1]
```diff
@@ -0,0 +1,1310 @@
...
+      updatedBudget.combined[year].medicaid += totalMedicaidCost;
+      updatedBudget.combined.total.medicaid += totalMedicaidCost;
+    }
+  }
+
+  return updatedBudget;
+};
+
+/**
+ * Sums the MMIS totals for each categories and total of all the categories
```
Comment (tbolt) [inline common/utils/budget.js:708]: categories -> category

**Review by tbolt — APPROVED** [review]
Nice work. Budget scans really well and documentation seems clear in general.

File: common/utils/utils.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  return Math.round(value * mult) / mult;
+};
+
+/**
+ * Round the values of an object to the given number to a given decimal place.
+ * @param {Object} o The object to round
+ * @param {Integer} digits The number of digits to round to, defaults to 2
+ * @returns {Object} the object with rounded values
+ */
+export const roundValues = (o, digits = 2) => {
```
[the remark this answers is not in the document]
Comment (thetif) [reply common/utils/utils.js:20]: changed it to roundObjectValues

File: common/utils/budget.js [hunk 3]
```diff
@@ -0,0 +1,1310 @@
...
+      combined: 0,
+      contractors: 0,
+      inHouse: 0
+    }
+  }
+});
+
+/**
+ * Creates a default Activity Totals data object
+ * @param {Array} years As a four-character year string (e.g., '2018')
```
[the remark this answers is not in the document]
Comment (thetif) [reply common/utils/budget.js:259]: changed them all to `The list of years in the APD`

File: common/utils/budget.js [hunk 4]
```diff
@@ -0,0 +1,1310 @@
...
+  statePersonnel,
+  keyPersonnel
+} = {}) => {
+  if (Array.isArray(statePersonnel)) {
+    const updatedStatePersonnel = [...statePersonnel];
+    // If this is the Program Administration activity, add the APD key
+    // personnel to it so they get included in downstream calculations.
+    // TODO: Don't rely on the activity name. Instead, this should always be
+    // activity index 0. However, that will require some tweaks to the tests.
+    if (name === 'Program Administration' && Array.isArray(keyPersonnel)) {
```
[the remark this answers is not in the document]
Comment (thetif) [reply common/utils/budget.js:342]: added more comments

**Review by mirano-darren — APPROVED** [review]
Looks good to me! Tested by comparing the seed data from main and this PR. Mainly compared the proposed budget section, which was identical to main.

**Review by jeromeleecms — COMMENTED** [review]
Proposed Budget doesn't seem to be populating properly

1) Created a new APD
2) Added a new Activity
3) Added a new State Staff (FFY 2022, $1 and FFY 2023, $4)
4) Review Proposed Budget, costs are not propagating to the summary tables

Example: 

Summary Budget Table (shows the total in the summary, but not the HIT cost)

image

Summary Budget Table (HIT Activities) - (doesn't show cost at all)

image

Quarterly Federal Share (no costs here):

image

Thoughts? Initially, the selection was missing the Federal/State split, but after adding that in, the numbers still did not populate.

**Comment by thetif** [comment]
strange, I'll look into it. Thanks for catching that, I'll add those to our regression tests, it seems like they should have caught it.

**Comment by thetif** [comment]
actually looking at this again, I think that's how it is supposed to work if you didn't specify a funding source

**Comment by thetif** [comment]
Jerome, I think it is acting as expected here are the same tables with the same input you gave above in staging. This is happening because the new activity does not have a funding source, so it's values get rolled into some of the totals but not anything that is counting values for each funding source.

Summary Budget by Activity
Screen Shot 2022-07-27 at 10 22 20 AM

Activity Breakdown
Screen Shot 2022-07-27 at 10 22 44 AM

Summary Budget Table
Screen Shot 2022-07-27 at 10 29 39 AM

Quarterly Federal Share
Screen Shot 2022-07-27 at 10 30 12 AM

**Review by jeromeleecms — APPROVED** [review]
Ok sounds like the admin check will address this, though users might know what is happening if they move the budget table before activating the admin check (but we can address that outside of this ticket). Otherwise, this looks good to go.
